pkg/standalone-utils/contracts/LPOracleFactoryBase.sol

```diff
@@ -0,0 +1,101 @@
+import { ILPOracleFactoryBase } from "@balancer-labs/v3-interfaces/contracts/standalone-utils/ILPOracleFactoryBase.sol";
+import { ILPOracleBase } from "@balancer-labs/v3-interfaces/contracts/standalone-utils/ILPOracleBase.sol";
+import { IBasePool } from "@balancer-labs/v3-interfaces/contracts/vault/IBasePool.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+
+import { InputHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/InputHelpers.sol";
+import { SingletonAuthentication } from "@balancer-labs/v3-vault/contracts/SingletonAuthentication.sol";
+
+/**
+ * @notice Base contract for deploying and managing pool oracles.
+ * @dev The factories that inherit this contract must implement the `_create` function, which deploys a different
+ * oracle, depending on the pool.
+ */
+abstract contract LPOracleFactoryBase is ILPOracleFactoryBase, SingletonAuthentication {
+    using Arrays for *;
+
+    uint256 internal _oracleVersion;
+    bool internal _isDisabled;
+
+    mapping(bytes32 oracleId => ILPOracleBase oracle) internal _oracles;
+    mapping(ILPOracleBase oracle => bool creationFlag) internal _isOracleFromFactory;
+
+    constructor(IVault vault, uint256 oracleVersion) SingletonAuthentication(vault) {
+        _oracleVersion = oracleVersion;
+    }
+
+    /// @inheritdoc ILPOracleFactoryBase
+    function create(IBasePool pool, AggregatorV3Interface[] memory feeds) external returns (ILPOracleBase oracle) {
+        _ensureEnabled();
+
+        bytes32 oracleId = _computeOracleId(pool, feeds);
+
+        if (address(_oracles[oracleId]) != address(0)) {
+            revert OracleAlreadyExists(pool, feeds, _oracles[oracleId]);
+        }
+
+        IVault vault = getVault();
+        IERC20[] memory tokens = vault.getPoolTokens(address(pool));
+
+        InputHelpers.ensureInputLengthMatch(tokens.length, feeds.length);
+
+        oracle = _create(vault, pool, feeds);
+        _oracles[oracleId] = oracle;
+        _isOracleFromFactory[oracle] = true;
+    }
+
+    /// @inheritdoc ILPOracleFactoryBase
+    function getOracle(
+        IBasePool pool,
+        AggregatorV3Interface[] memory feeds
+    ) external view returns (ILPOracleBase oracle) {
+        bytes32 oracleId = _computeOracleId(pool, feeds);
+        oracle = ILPOracleBase(address(_oracles[oracleId]));
+    }
+
+    /// @inheritdoc ILPOracleFactoryBase
+    function isOracleFromFactory(ILPOracleBase oracle) external view returns (bool success) {
+        success = _isOracleFromFactory[oracle];
+    }
+
+    /// @inheritdoc ILPOracleFactoryBase
+    function disable() external authenticate {
+        _ensureEnabled();
+
+        _isDisabled = true;
+        emit OracleFactoryDisabled();
+    }
+
+    function _computeOracleId(IBasePool pool, AggregatorV3Interface[] memory feeds) internal pure returns (bytes32) {
+        address[] memory feedAddresses = asAddress(feeds);
+        return keccak256(abi.encode(pool, feedAddresses));
```

Comment:
Simplified this a bit to avoid unnecessary copies.

Also skipped sorting the feeds. You can create an oracle with the fees in the wrong order, and therefore need to create it again. If we sort it here, two oracles with the same feeds but in different order would be considered as exactly the same, whereas that's not the case really. Each feed corresponds to a token, in whatever order they come.